### ISSUE TITLE
build method is triggered multiple times when Keyboard show/hide inside `showModalBottomSheet`

### ISSUE THREAD

**Issue by pickledbrill** [flutter/flutter#130560]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

When a `TextFormField` is focused inside `showModalBottomSheet`, the `build` method is being called several times when the keyboard shows or hide.

### Expected results

From my understanding, the `build` method is expected to be called only once. From the provided code sample, the `build method is called...` should only appear once.

### Actual results

`build method is called...` appears over 10 times when the keyboard shows or hide.

### Code sample

<details><summary>Code sample</summary>

```dart
class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    print('build method is called...');

    return Scaffold(
      resizeToAvoidBottomInset: false,
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: ElevatedButton(
          onPressed: () {
            showModalBottomSheet(
                isDismissible: true,
                isScrollControlled: true,
                context: context,
                builder: (ctx) {
                  return SizedBox(
                    height: MediaQuery.of(context).size.height * 0.5,
                    child: TextFormField(
                      decoration: const InputDecoration(
                          prefixIcon: Icon(Icons.search_outlined), hintText: 'input', border: InputBorder.none),
                    ),
                  );
                });
          },
          child: Text('Show ModalBottomSheet'),
        ),
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/52510584/9306983a-efe4-4776-ab96-ccbc8432eded

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
flutter doctor -v
[✓] Flutter (Channel master, 3.10.0-10.0.pre.17, on macOS 13.4 22F66 darwin-arm64, locale zh-Hans-CN)
    • Flutter version 3.10.0-10.0.pre.17 on channel master at /Users/billchen/Flutter/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b04efe4ecf (3 months ago), 2023-04-20 01:31:24 -0400
    • Engine revision 2db85cbe6a
    • Dart version 3.1.0 (build 3.1.0-22.0.dev)
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0-rc3)
    • Android SDK at /Users/billchen/Library/Android/sdk
    • Platform android-33, build-tools 34.0.0-rc3
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] IntelliJ IDEA Ultimate Edition (version 2021.2.4)
    • IntelliJ at /Applications/IntelliJ IDEA 2.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.80.1)
    • VS Code at /Applications/Visual Studio Code 2.app/Contents
    • Flutter extension version 3.68.0

[✓] Connected device (3 available)
    • sdk gphone64 arm64 (mobile) • emulator-5554 • android-arm64  • Android 13 (API 33) (emulator)
    • macOS (desktop)             • macos         • darwin-arm64   • macOS 13.4 22F66 darwin-arm64
    • Chrome (web)                • chrome        • web-javascript • Google Chrome 114.0.5735.198

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by dam-ease** [comment]
Hi @pickledbrill. Thanks for filing this issue. This seems to be caused by the use of `MediaQuery.of(context)` (removing this, the build method doesn't get triggered multiple times) and there was a fix to that effect https://github.com/flutter/flutter/pull/114459 but testing on the latest `master` and `stable` channels, it seems to have not been fixed. 
You can try this suggested solution https://github.com/flutter/flutter/issues/105184#issuecomment-1566737581 if it works in your use case. I'd be closing this as a duplicate of https://github.com/flutter/flutter/issues/105184. Kindly comment and upvote this issue.
If you disagree, please write in the comments and I will reopen it.
Thank you
